File: ios/ptchampion/Grading/Grading/PullupGrader.swift
import Foundation
import Vision
import CoreGraphics

class PullupGrader: ExerciseGraderProtocol {

    private enum PullupPhase {
        case down // Arms fully extended hanging
        case up   // Chin above the bar
        case starting
        case invalid
        case between // Transitioning
    }

    private var currentState: PullupPhase = .starting
    private var feedback: String = "Hang from bar, arms extended."
    private var repCount: Int = 0

    // Configuration Thresholds (tune based on testing!)
    private let elbowAngleDownMin: CGFloat = 160.0  // Min angle to be considered fully extended ('down') (Aligned w/ Android)
    // Removed elbowAngleUpMax - UP state depends on chin height (Aligned w/ Android)
    // Vertical distance threshold: Nose Y must be *above* Wrist Y by this amount (normalized)
    // Note: In Vision coordinates (0,0 is bottom-left), smaller Y is higher.
    private let chinAboveBarMinYDiff: CGFloat = 0.05 // Aligned w/ Android
    // Max allowed vertical hip movement relative to shoulder height (normalized) - basic kipping check
    private let kippingMaxHipYTravel: CGFloat = 0.10 // Aligned w/ Android
    private let requiredConfidence: Float = 0.6 // Aligned w/ Android
    // Minimum bend required during UP phase for rep check
    private let elbowAngleUpRepCheckMax: CGFloat = 90.0 // Aligned w/ Android MIN_BEND_THRESHOLD

    // Additional thresholds that were referenced but not previously declared
    // Relative scale (multiplicative) factors applied to shoulder Y reference when classifying states.
    // Tuned empirically; feel free to adjust during real‑world testing.
    private let chinHeightThreshold: CGFloat = 0.05          // wrists this much (or less) *above* shoulders counts as "up"
    private let elbowShoulderDeadHangThreshold: CGFloat = 0.20 // wrists this much (or more) *below* shoulders counts as "down"

    // Form‑tracking helpers that were referenced but never declared
    private var formIssues: [String] = []
    private var formScore: Int = 100

    // State tracking for rep evaluation & stability
    private var maxElbowAngleThisRep: CGFloat = 0.0      // Track max angle for extension check
    private var minElbowAngleThisRep: CGFloat = 180.0    // Track min angle during UP phase for rep check
    private var chinWasAboveBarThisRep: Bool = false   // Track if chin cleared bar during UP phase
    private var startingHipY: CGFloat? = nil         // Initial hip Y for kipping check
    private var minHipYThisRep: CGFloat = 1.0          // Track min/max hip Y during rep
    private var maxHipYThisRep: CGFloat = 0.0
    private var stableFrameCounter: Int = 0
    private let requiredStableFrames: Int = 3 // Frames needed to confirm state change (Aligned w/ Android)
    private var repInProgress: Bool = false // Flag when moving from DOWN state

    var currentPhaseDescription: String {
        switch currentState {
        case .down: return "Down (Hang)"
        case .up: return "Up (Chin Above)"
        case .starting: return "Ready"
        case .invalid: return "Invalid Pose"
        case .between: return "Moving"
        }
    }

    func resetState() {
        currentState = .starting
        feedback = "Hang from bar, arms extended."
        repCount = 0
        resetRepTrackingState()
        stableFrameCounter = 0
        repInProgress = false
        startingHipY = nil
        print("PullupGrader: State reset.")
    }

    private func resetRepTrackingState() {
        maxElbowAngleThisRep = 0.0
        minElbowAngleThisRep = 180.0
        chinWasAboveBarThisRep = false
        // Don't reset startingHipY here, only on first valid frame or full reset
        minHipYThisRep = 1.0
        maxHipYThisRep = 0.0
    }

    func gradePose(body: DetectedBody) -> GradingResult {
        feedback = "" // Reset feedback
        formIssues.removeAll()

        // Keep a snapshot of the state before analysing this frame
        let previousState = currentState
        // Default grading outcome; will be updated below
        var gradingResult: GradingResult = .noChange

        // 1. Check Required Joint Confidence
        let keyJoints: [VNHumanBodyPoseObservation.JointName] = [
            .leftShoulder, .rightShoulder,
            .leftElbow, .rightElbow,
            .leftWrist, .rightWrist,
            .leftHip, .rightHip, // Needed for kipping check
            .nose // Needed for chin height check
        ]
        var missingJoints: [String] = []
        for jointName in keyJoints {
            guard let point = body.point(jointName), point.confidence >= requiredConfidence else {
                missingJoints.append(jointName.rawValue.rawValue)
                continue // Check all missing joints
            }
        }
        if !missingJoints.isEmpty {
            updateState(to: .invalid, stable: false)
            feedback = "Cannot see clearly: \(missingJoints.joined(separator: ", "))"
            return .invalidPose(reason: feedback)
        }

        // Extract validated points
        let leftShoulder = body.point(.leftShoulder)!
        let rightShoulder = body.point(.rightShoulder)!
        let leftElbow = body.point(.leftElbow)!
        let rightElbow = body.point(.rightElbow)!
        let leftWrist = body.point(.leftWrist)!
        let rightWrist = body.point(.rightWrist)!
        let nose = body.point(.nose)!
        let leftHip = body.point(.leftHip)!
        let rightHip = body.point(.rightHip)!

        // 2. Calculate Key Angles & Positions
        let leftElbowAngle = calculateAngle(point1: leftShoulder.location, centerPoint: leftElbow.location, point2: leftWrist.location)
        let rightElbowAngle = calculateAngle(point1: rightShoulder.location, centerPoint: rightElbow.location, point2: rightWrist.location)
        let avgElbowAngle = averageAngle(leftElbowAngle, rightElbowAngle) ?? 0.0 // Default bent

        let avgWristY = (leftWrist.location.y + rightWrist.location.y) / 2.0
        let noseY = nose.location.y
        let chinIsAboveBar = noseY < (avgWristY - chinAboveBarMinYDiff) // Smaller Y is higher

        let avgHipY = (leftHip.location.y + rightHip.location.y) / 2.0

        // Set starting hip position on first valid frame in starting/down state
        if startingHipY == nil && (currentState == .starting || currentState == .down) && avgElbowAngle >= elbowAngleDownMin * 0.95 {
             startingHipY = avgHipY
        }

        // Establish a shoulder‑level reference for relative Y‑position checks
        let shoulderReferenceY = (leftShoulder.location.y + rightShoulder.location.y) / 2.0

        // 1. Determine potential state.
        // We consider "up" when the chin clears the bar (chinIsAboveBar).
        // We consider "down" when the elbows are almost fully extended.
        var potentialState: PullupPhase = currentState
        if chinIsAboveBar {
            potentialState = .up
        } else if avgElbowAngle >= elbowAngleDownMin {
            potentialState = .down
        } else {
            potentialState = .between
        }

        currentState = potentialState

        // 2. Check for Rep Completion: Transitioned from Down to Up
        if previousState == .down && currentState == .up {
            repCount += 1
            gradingResult = .repCompleted
            feedback = "Rep Counted! (\(repCount))"
        } else {
            // Provide feedback based on state if no rep was just counted
            switch currentState {
            case .up: feedback = "Lower yourself."
            case .down: feedback = "Pull up."
            case .starting: feedback = "Begin pull-ups."
            case .between: feedback = "Keep moving."
            case .invalid: feedback = "Fix pose."
            }
            // Use if-case to check GradingResult without needing Equatable conformance
            if case .inProgress(let phase) = gradingResult {
                 print("Still in progress: \(phase)") // Example placeholder
            } else if case .noChange = gradingResult {
                 gradingResult = .inProgress(phase: currentPhaseDescription) // Update if still just in progress
            }

        }
        // Update form score (simple pass/fail for now)
        formScore = formIssues.isEmpty ? 100 : 0 // Penalize fully for any form issue

        // Add specific form feedback if issues exist
        if !formIssues.isEmpty {
            feedback = formIssues.joined(separator: " ")
            // Use if-case to check GradingResult without needing Equatable conformance
            if case .inProgress(let phase) = gradingResult {
                 print("Still in progress with form issues: \(phase)") // Example placeholder
            } else if case .noChange = gradingResult {
                 gradingResult = .incorrectForm(feedback: feedback)
            }

        }

        // Start tracking rep if moving from DOWN state
        if previousState == .down && currentState != .down && !repInProgress {
            repInProgress = true
            resetRepTrackingState() // Reset tracking for the new rep attempt
             if startingHipY == nil { startingHipY = avgHipY } // Capture hip starting Y if missed
        }

        // Track angles and positions during the rep
        if repInProgress {
            maxElbowAngleThisRep = max(maxElbowAngleThisRep, avgElbowAngle)
            minElbowAngleThisRep = min(minElbowAngleThisRep, avgElbowAngle) // Track min angle too
            if chinIsAboveBar { chinWasAboveBarThisRep = true }
            minHipYThisRep = min(minHipYThisRep, avgHipY)
            maxHipYThisRep = max(maxHipYThisRep, avgHipY)
        }

        // Check for Rep Completion (Transition UP -> DOWN)
        if previousState == .up && currentState == .down && repInProgress {
            var repFormIssues: [String] = []

            // a) Check Full Extension (using max angle achieved during rep)
            if maxElbowAngleThisRep < elbowAngleDownMin {
                repFormIssues.append("Extend arms fully at bottom.")
            }
            // b) Check Chin Over Bar (must have been true at some point during UP)
            if !chinWasAboveBarThisRep {
                 repFormIssues.append("Chin did not clear bar.")
            }
            // c) Check Kipping (Hip Y travel)
            let hipTravel = maxHipYThisRep - minHipYThisRep
            // Estimate body height (shoulder to hip) for normalization - crude approximation
            let bodyHeightEstimate = abs(leftShoulder.location.y - leftHip.location.y) + abs(rightShoulder.location.y - rightHip.location.y) / 2.0
            if startingHipY != nil && bodyHeightEstimate > 0.1 && (hipTravel / bodyHeightEstimate) > kippingMaxHipYTravel {
                 repFormIssues.append("Excessive hip movement (kipping).")
            }
            // d) Check Minimum Elbow Bend during Up phase (Aligned w/ Android)
            if minElbowAngleThisRep > elbowAngleUpRepCheckMax {
                 repFormIssues.append("Arms not bent enough at top.")
            }

            if repFormIssues.isEmpty {
                repCount += 1
                feedback = "Rep Counted! (\(repCount))"
                gradingResult = .repCompleted
            } else {
                feedback = repFormIssues.joined(separator: " ")
                gradingResult = .incorrectForm(feedback: feedback)
            }
            // Reset tracking for the next rep attempt, mark rep as no longer in progress
            repInProgress = false
            resetRepTrackingState()
            // Don't reset startingHipY unless fully resetting

        } else {
            // Provide general feedback if no rep completed
            gradingResult = .inProgress(phase: currentPhaseDescription)
             if feedback.isEmpty { // Avoid overwriting specific form issue feedback
                 switch currentState {
                 case .down: feedback = "Pull up."
                 case .up: feedback = "Lower down slowly."
                 case .starting: feedback = "Begin when ready."
                 case .between:
                     if previousState == .down { feedback = "Pulling up..." }
                     else if previousState == .up { feedback = "Lowering..." }
                     else { feedback = "Keep moving."}
                 case .invalid: feedback = "Fix pose."
                 }
            }
        }

         // Immediate feedback on current position if needed
         if currentState == .down && avgElbowAngle < elbowAngleDownMin {
             // Use if-case for comparison
             if case .inProgress = gradingResult {
                 feedback = "Extend arms fully."
                 gradingResult = .incorrectForm(feedback: feedback)
             } else if case .noChange = gradingResult {
                 feedback = "Extend arms fully."
                 gradingResult = .incorrectForm(feedback: feedback)
             }
         } else if currentState == .up && !chinIsAboveBar {
             // Use if-case for comparison
             if case .inProgress = gradingResult {
                 feedback = "Pull higher!"
                 gradingResult = .incorrectForm(feedback: feedback)
             } else if case .noChange = gradingResult {
                 feedback = "Pull higher!"
                 gradingResult = .incorrectForm(feedback: feedback)
             }
         }

        return gradingResult
    }

    // Update state only if stable for enough frames
    private func updateState(to newState: PullupPhase, stable: Bool = true) {
        if newState == currentState {
            if stable { stableFrameCounter += 1 }
        } else {
            // Only change if required frames met OR the new state is invalid
            // Also allow immediate transition *out* of invalid state
            if !stable || stableFrameCounter >= requiredStableFrames || newState == .invalid || currentState == .invalid {
                 // Reset counter only if the state *actually* changes
                 if currentState != newState { stableFrameCounter = 0 }
                 currentState = newState
            } // else: keep current state until stability threshold is met
        }
    }

    // Helper to average angles
     private func averageAngle(_ angle1: CGFloat?, _ angle2: CGFloat?) -> CGFloat? {
         switch (angle1, angle2) {
         case (.some(let a1), .some(let a2)): return (a1 + a2) / 2.0
         case (.some(let a1), .none): return a1
         case (.none, .some(let a2)): return a2
         case (.none, .none): return nil
         }
     }
     
     func calculateFinalScore() -> Double? {
         // Simple score implementation based on form quality
         return Double(formScore)
     }
}

File: web/src/grading/PullupGrader.ts
/**
 * PullupGrader.ts
 * 
 * Implements pull-up specific grading logic.
 * This encapsulates the state machine and form evaluation criteria for pull-ups.
 */

import { NormalizedLandmark } from '@mediapipe/tasks-vision';
import { BaseExerciseGrader, GradingResult } from './ExerciseGrader';
import { PoseLandmarkIndex } from '../services/PoseLandmarkIndex';

/**
 * Constants for pull-up form evaluation
 */
export const PULLUP_CONSTANTS = {
  // Angle for fully extended arms at bottom
  THRESHOLD_ELBOW_ANGLE_DOWN: 160,
  
  // Threshold for chin (nose landmark) being above wrists
  // Positive value means nose needs to be *higher* than wrists (smaller y-coordinate)
  CHIN_OVER_BAR_VERTICAL_THRESHOLD: 0.05,
  
  // Threshold to detect excessive kipping (vertical hip movement relative to shoulders)
  KIPPING_VERTICAL_THRESHOLD: 0.15,
  
  // Visibility threshold for landmarks
  VISIBILITY_THRESHOLD: 0.6
};

/**
 * Required landmarks for pull-up tracking
 */
const REQUIRED_PULLUP_LANDMARKS = [
  PoseLandmarkIndex.NOSE,
  PoseLandmarkIndex.LEFT_SHOULDER,
  PoseLandmarkIndex.RIGHT_SHOULDER,
  PoseLandmarkIndex.LEFT_ELBOW,
  PoseLandmarkIndex.RIGHT_ELBOW,
  PoseLandmarkIndex.LEFT_WRIST,
  PoseLandmarkIndex.RIGHT_WRIST,
  PoseLandmarkIndex.LEFT_HIP,
  PoseLandmarkIndex.RIGHT_HIP
];

/**
 * Pull-up grader that implements the exercise state machine and form evaluation
 */
export class PullupGrader extends BaseExerciseGrader {
  // Track form faults during the current rep
  private currentFormFault: string | null = null;
  
  // Store initial positions for checking kipping
  private initialHipY: number | null = null;
  private initialShoulderY: number | null = null;
  
  constructor() {
    super('PULLUP');
  }
  
  /**
   * Process a new frame of pose landmarks and update the pull-up state
   * @param landmarks Array of pose landmarks from MediaPipe
   * @returns GradingResult with updated state and rep information
   */
  processPose(landmarks: NormalizedLandmark[]): GradingResult {
    // Default result with no rep increment and no form fault
    let result: GradingResult = {
      state: this.state,
      repIncrement: 0,
      hasFormFault: false,
      formScore: 100 // Start with perfect form, deduct as needed
    };
    
    // Check if all required landmarks are visible
    if (!this.areLandmarksVisible(landmarks, REQUIRED_PULLUP_LANDMARKS, PULLUP_CONSTANTS.VISIBILITY_THRESHOLD)) {
      // Not enough landmarks visible for reliable tracking
      if (this.state !== 'start') {
        // Only reset state and show message if we were in the middle of tracking
        this.state = 'start';
        this.currentFormFault = null;
        this.initialHipY = null;
        this.initialShoulderY = null;
        
        result.state = 'start';
        result.formFault = "Ensure full body is visible";
        result.hasFormFault = true;
      }
      return result;
    }
    
    // Extract relevant landmarks
    const nose = landmarks[PoseLandmarkIndex.NOSE];
    const leftShoulder = landmarks[PoseLandmarkIndex.LEFT_SHOULDER];
    const rightShoulder = landmarks[PoseLandmarkIndex.RIGHT_SHOULDER];
    const leftElbow = landmarks[PoseLandmarkIndex.LEFT_ELBOW];
    const rightElbow = landmarks[PoseLandmarkIndex.RIGHT_ELBOW];
    const leftWrist = landmarks[PoseLandmarkIndex.LEFT_WRIST];
    const rightWrist = landmarks[PoseLandmarkIndex.RIGHT_WRIST];
    const leftHip = landmarks[PoseLandmarkIndex.LEFT_HIP];
    const rightHip = landmarks[PoseLandmarkIndex.RIGHT_HIP];
    
    // Calculate angles and positions
    const leftElbowAngle = this.calculateAngle(leftShoulder, leftElbow, leftWrist);
    const rightElbowAngle = this.calculateAngle(rightShoulder, rightElbow, rightWrist);
    const elbowAngle = (leftElbowAngle + rightElbowAngle) / 2; // Average elbow angle
    
    // Calculate vertical positions (y-coordinate, smaller is higher on screen)
    const noseY = nose.y;
    const avgWristY = (leftWrist.y + rightWrist.y) / 2;
    const avgShoulderY = (leftShoulder.y + rightShoulder.y) / 2;
    const avgHipY = (leftHip.y + rightHip.y) / 2;
    
    // Conditions for state transitions
    const isArmsExtended = elbowAngle >= PULLUP_CONSTANTS.THRESHOLD_ELBOW_ANGLE_DOWN;
    // Check if nose is significantly higher than wrists
    const isChinOverBar = noseY < avgWristY - PULLUP_CONSTANTS.CHIN_OVER_BAR_VERTICAL_THRESHOLD;
    
    // State machine transitions
    let kippingFaultDetected = false;
    
    switch (this.state) {
      case 'start':
        // Waiting to reach the bottom hang position (arms extended)
        if (isArmsExtended) {
          this.state = 'down';
          this.currentFormFault = null; // Reset fault flag for the new rep attempt
          
          // Record initial positions for kipping check
          this.initialHipY = avgHipY;
          this.initialShoulderY = avgShoulderY;
          
          console.log("State -> DOWN (Arms Extended)");
          result.state = 'down';
        }
        break;
        
      case 'down':
        // At the bottom, waiting for upward movement AND chin over bar
        if (isChinOverBar) {
          // Check for excessive kipping during the upward movement
          if (this.initialHipY !== null && this.initialShoulderY !== null) {
            const currentRelativeHipY = avgHipY - avgShoulderY;
            const initialRelativeHipY = this.initialHipY - this.initialShoulderY;
            const verticalHipDisplacement = initialRelativeHipY - currentRelativeHipY;
            
            if (verticalHipDisplacement > PULLUP_CONSTANTS.KIPPING_VERTICAL_THRESHOLD) {
              this.currentFormFault = "Excessive Kipping Detected!";
              kippingFaultDetected = true;
              console.warn(`Form Fault: ${this.currentFormFault} Displacement: ${verticalHipDisplacement.toFixed(3)}`);
              result.hasFormFault = true;
              result.formScore = 70; // Deduct more points for kipping
            }
          }
          
          // Transition to 'up' regardless of kipping for state flow, fault flag handles counting
          this.state = 'up';
          console.log(`State -> UP (Chin Over Bar: ${isChinOverBar}, Kipping Fault: ${kippingFaultDetected})`);
          result.state = 'up';
          
          if (this.currentFormFault) {
            result.formFault = this.currentFormFault;
          }
        }
        break;
        
      case 'up':
        // At the top, waiting for downward movement (arms extending)
        if (isArmsExtended) {
          // Successfully returned to bottom hang position
          if (!result.hasFormFault && !kippingFaultDetected) {
            // Only count if no form fault was detected during the rep
            result.repIncrement = 1; // Count the rep!
            console.log("State -> DOWN (Rep Counted)");
          } else {
            console.log("State -> DOWN (Rep Not Counted due to Form Fault)");
            // Give feedback that the previous rep was invalid
            if (!this.currentFormFault) {
              this.currentFormFault = "Invalid Rep - Check Form";
            }
            result.formFault = this.currentFormFault;
          }
          
          this.state = 'down'; // Go back to down state, ready for next rep
          result.state = 'down';
          this.currentFormFault = null; // Reset fault flag for the NEW rep cycle
          
          // Record new initial positions
          this.initialHipY = avgHipY;
          this.initialShoulderY = avgShoulderY;
        }
        break;
    }
    
    // Update any form fault status from this frame into the result
    if (this.currentFormFault && !result.formFault) {
      result.formFault = this.currentFormFault;
      result.hasFormFault = true;
    }
    
    return result;
  }
  
  /**
   * Reset the pull-up grader to its initial state
   */
  reset(): void {
    super.reset();
    this.currentFormFault = null;
    this.initialHipY = null;
    this.initialShoulderY = null;
  }
}

export default PullupGrader; 